pkg/controller/elasticsearch/driver/upgrade_forced.go

```diff
@@ -0,0 +1,52 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package driver
+
+import (
+	corev1 "k8s.io/api/core/v1"
+
+	"github.com/elastic/cloud-on-k8s/pkg/utils/k8s"
+)
+
+// maybeForceUpgrade may attempt a forced upgrade of all podsToUpgrade if allowed to,
+// in order to unlock situations where the reconciliation may otherwise be stuck
+// (eg. no cluster formed, all nodes have a bad spec).
+func (d *defaultDriver) maybeForceUpgrade(actualPods []corev1.Pod, podsToUpgrade []corev1.Pod) (attempted bool, err error) {
+	attempted = false
+	if len(podsToUpgrade) == 0 {
+		return attempted, nil
+	}
+	if !shouldForceUpgrade(actualPods) {
+		return attempted, nil
+	}
+
+	attempted = true
+	log.Info("Performing a forced rolling upgrade since no Pod is ready",
+		"namespace", d.ES.Namespace, "es_name", d.ES.Name,
+		"pod_count", len(podsToUpgrade),
+	)
+	for _, pod := range podsToUpgrade {
```

Comment:
I was tempted to respect the change budget here (ie. if `maxUnavailable=1`, only restart one Pod at a time). The problem is the restarted Pod would become Ready (eg. misconfiguration fixed), but the cluster itself would probably not. If we have 1/15 Pods ready with missing Masters, no chance the cluster turns green. And then we would be stuck again, since we'd wait for the cluster to turn green, which would never happen.

Proceeding with all Pods at once is a way to handle that. However if the operator happens to crash in the middle of al pods deletion, there's a risk we end up in the situation above. I don't think that's a corner case worth being handled explicitly.

Reply:
I wonder if the heuristics to detect a situation where we want to force an upgrade needs to be a little bit smarter. E.g. we have some status information on the pods that would allow us to detect a bootloop with much more precision then just looking at all pods ready state. While just looking at the ready state is more general I feel like it is also susceptible to such edge cases because the decision is made based on a momentary state (all pods not ready) while the restart count seems more resilient?

Reply:
* That makes sense! Relying on `restartCount` (+ non-ready) to detect bootlooping Pods sounds indeed better.

* Another common issue is Pods that stay Pending forever (see the corresponding e2e test). I guess to fix that situation we could also filter on "all Pods pending for more than X seconds".

* A third one: I'm wondering if we may also have a situation where Pods are Running, but not Ready. ie. Elasticsearch is started but not responding for whatever reason (misconfiguration probably). Not sure how to produce (if possible) that behaviour though.